examples/solids/src/cloptions.c

```diff
@@ -82,12 +97,46 @@ PetscErrorCode ProcessCommandLineOptions(MPI_Comm comm, AppCtx appCtx) {
                               "Face IDs to apply incremental Dirichlet BC",
                               NULL, appCtx->bcClampFaces, &appCtx->bcClampCount,
                               NULL); CHKERRQ(ierr);
-
-  appCtx->bcClampMax = -1;
-  ierr = PetscOptionsScalar("-bc_clamp_max",
-                            "Maximum value to displace clamped boundary",
-                            NULL, appCtx->bcClampMax, &appCtx->bcClampMax,
-                            NULL); CHKERRQ(ierr);
+  // Set vector for each clamped BC
+  for (PetscInt i = 0; i < appCtx->bcClampCount; i++) {
+    // Translation vector
+    char optionName[25];
+    snprintf(optionName, sizeof optionName, "-bc_clamp_%d_translate",
+             appCtx->bcClampFaces[i]);
+    maxn = 3;
+
+    appCtx->bcClampTranslate[i] = PETSC_FALSE;
+    ierr = PetscOptionsScalarArray(optionName,
+                                   "Vector to translate clamped end by", NULL,
+                                   appCtx->bcClampMax[i], &maxn,
+                                   &appCtx->bcClampTranslate[i]);
+    CHKERRQ(ierr);
+
+    // Rotation vector
+    if (!appCtx->bcClampTranslate[i]) {
+      PetscBool bcFlag = PETSC_FALSE;
+      maxn = 4;
+      snprintf(optionName, sizeof optionName, "-bc_clamp_%d_rotate",
+               appCtx->bcClampFaces[i]);
+      ierr = PetscOptionsScalarArray(optionName,
+                                     "Rodrigues vector to rotate clamped end by, in radians",
+                                     NULL, appCtx->bcClampMax[i], &maxn,
+                                     &bcFlag); CHKERRQ(ierr);
+
+      // Normalize
+      PetscScalar norm = sqrt(appCtx->bcClampMax[i][0]*appCtx->bcClampMax[i][0]+
+                              appCtx->bcClampMax[i][1]*appCtx->bcClampMax[i][1]+
+                              appCtx->bcClampMax[i][2]*appCtx->bcClampMax[i][2]);
+      for (PetscInt j = 0; j < 3; j++)
+        appCtx->bcClampMax[i][j] /= norm;
+
+      // Error if neither were set
+      if (!bcFlag) {
+        SETERRQ(PETSC_COMM_SELF, PETSC_ERR_SUP,
+                "-bc_clamp additional option needed");
```

Comment:
How about `-bc_clamp` defaults to zero translation and rotation. That way we just have one option (get rid of `-bc_zero`).

We can (in this PR or another) add `-bc_clamp_%d_slip 1,0,0` to specify a vector normal to the plane on which we allow free slip (no friction or contact). This would most physically be applied in the current configuration, but that creates a nonlinearity in the boundary condition and defining it in the reference configuration is sufficient in this case (due to the nature of a clamped planar boundary).

Reply:
Also agreed. It might make sense to get these changes into master and then add no slip?

Reply:
Sure, can you clean up the affine matrix formulation in this PR? Happy to merge.